src/Models/HydrostaticFreeSurfaceModels/mg_implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,116 @@
+    "The multigrid solver"
+    multigrid_solver :: S
+    "The vertically-integrated lateral areas"
+    vertically_integrated_lateral_areas :: V
+    "The previous time step"
+    previous_Δt :: F
+    "The right hand side of the free surface evolution equation"
+    right_hand_side :: R
+end
+
+architecture(solver::MGImplicitFreeSurfaceSolver) =
+    architecture(solver.multigrid_solver)
+
+"""
+    PCGImplicitFreeSurfaceSolver(grid, settings)
+
+Return a solver based on a preconditioned conjugate gradient method for
+the elliptic equation
+    
+```math
+[∇ ⋅ H ∇ - 1 / (g Δt²)] ηⁿ⁺¹ = (∇ʰ ⋅ Q★ - ηⁿ / Δt) / (g Δt)
+```
+
+representing an implicit time discretization of the linear free surface evolution equation
+for a fluid with variable depth `H`, horizontal areas `Az`, barotropic volume flux `Q★`, time
+step `Δt`, gravitational acceleration `g`, and free surface at time-step `n` `ηⁿ`.
+"""
+function MGImplicitFreeSurfaceSolver(grid::AbstractGrid, 
+                                    settings, 
+                                    gravitational_acceleration=nothing, 
+                                    placeholder_timestep = -1.0)
+
+    # Initialize vertically integrated lateral face areas
+    ∫ᶻ_Axᶠᶜᶜ = Field{Face, Center, Nothing}(with_halo((3, 3, 1), grid))
+    ∫ᶻ_Ayᶜᶠᶜ = Field{Center, Face, Nothing}(with_halo((3, 3, 1), grid))
+
+    vertically_integrated_lateral_areas = (xᶠᶜᶜ = ∫ᶻ_Axᶠᶜᶜ, yᶜᶠᶜ = ∫ᶻ_Ayᶜᶠᶜ)
+
+    compute_vertically_integrated_lateral_areas!(vertically_integrated_lateral_areas)
+    fill_halo_regions!(vertically_integrated_lateral_areas)
+    
+    # Set some defaults
+    # TODO edit these to be maximum_iterations and tolerance
+    # settings = Dict{Symbol, Any}(settings)
+    # settings[:maxiter] = get(settings, :maxiter, grid.Nx * grid.Ny)
+    # settings[:reltol] = get(settings, :reltol, min(1e-7, 10 * sqrt(eps(eltype(grid)))))
+
+    right_hand_side = Field{Center, Center, Nothing}(grid)
+
+    solver = MultigridSolver(grid, implicit_free_surface_linear_operation!, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ, gravitational_acceleration, nothing;)
+                                                #    settings...)
+
+    return MGImplicitFreeSurfaceSolver(solver, vertically_integrated_lateral_areas, placeholder_timestep, right_hand_side)
+end
+
+build_implicit_step_solver(::Val{:Multigrid}, grid, settings, gravitational_acceleration) =
+    MGImplicitFreeSurfaceSolver(grid, settings, gravitational_acceleration)
+
+#####
+##### Solve...
+#####
+
+function solve!(η, implicit_free_surface_solver::MGImplicitFreeSurfaceSolver, rhs, g, Δt)
+    #event = explicit_ab2_step_free_surface!(free_surface, model, Δt, χ)
+    #wait(device(model.architecture), event)
+    solver = implicit_free_surface_solver.multigrid_solver
+
+    # If Δt varies then construct matrix
+    if Δt != implicit_free_surface_solver.previous_Δt
+        ∫ᶻA = implicit_free_surface_solver.vertically_integrated_lateral_areas
```

Comment:
this does not change, right? so it should be computed upon initialization and then never again, right?

Reply:
Do you mean `∫ᶻA`? This is computed upon initialization, line 90 is just giving it a helpful name - or am I missing something?

Reply:
Yes, I was only referring to `∫ᶻA`